### ISSUE TITLE
`missing_asserts_for_indexing` fires for indexing inside `if` after `assert!`

### ISSUE THREAD

**Issue by jwodder** [rust-lang/rust-clippy#11856]
### Summary

Look at the code below; it says it better than words can.

### Lint Name

missing_asserts_for_indexing

### Reproducer

I tried this code:

```rust
#[deny(clippy::missing_asserts_for_indexing)]
pub fn foo(values: &[i32]) -> usize {
    let mut ascending = Vec::new();
    for w in values.windows(2) {
        assert!(w.len() > 1);
        if w[0] < w[1] {
            ascending.push((w[0], w[1]));
        } else {
            ascending.push((w[1], w[0]));
        }
    }
    ascending.len()
}
```

I expected to see this happen: No errors.

Instead, this happened:

```
    Checking multi-index v0.1.0 (/Users/jwodder/work/dev/tmp/multi-index)
error: indexing into a slice multiple times without an `assert`
 --> src/lib.rs:7:29
  |
7 |             ascending.push((w[0], w[1]));
  |                             ^^^^^^^^^^
  |
  = help: consider asserting the length before indexing: `assert!(w.len() > 1);`
note: slice indexed here
 --> src/lib.rs:7:29
  |
7 |             ascending.push((w[0], w[1]));
  |                             ^^^^
note: slice indexed here
 --> src/lib.rs:7:35
  |
7 |             ascending.push((w[0], w[1]));
  |                                   ^^^^
  = note: asserting the length before indexing will elide bounds checks
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#missing_asserts_for_indexing
note: the lint level is defined here
 --> src/lib.rs:1:8
  |
1 | #[deny(clippy::missing_asserts_for_indexing)]
  |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: indexing into a slice multiple times without an `assert`
 --> src/lib.rs:9:29
  |
9 |             ascending.push((w[1], w[0]));
  |                             ^^^^^^^^^^
  |
  = help: consider asserting the length before indexing: `assert!(w.len() > 1);`
note: slice indexed here
 --> src/lib.rs:9:29
  |
9 |             ascending.push((w[1], w[0]));
  |                             ^^^^
note: slice indexed here
 --> src/lib.rs:9:35
  |
9 |             ascending.push((w[1], w[0]));
  |                                   ^^^^
  = note: asserting the length before indexing will elide bounds checks
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#missing_asserts_for_indexing

error: could not compile `multi-index` (lib) due to 2 previous errors
warning: build failed, waiting for other jobs to finish...
error: could not compile `multi-index` (lib test) due to 2 previous errors
```

### Version

```text
rustc 1.74.0 (79e9716c9 2023-11-13)
binary: rustc
commit-hash: 79e9716c980570bfd1f666e3b16ac583f0168962
commit-date: 2023-11-13
host: x86_64-apple-darwin
release: 1.74.0
LLVM version: 17.0.4
```

**Comment by jwodder** [comment]
Uh, I meant to file this as a false positive.  I guess I must have clicked on "false negative" instead.
